tools/autograd/gen_variable_type.py

```diff
@@ -160,146 +194,220 @@ def write(dirname, name, template, env):
+    seen = set()
+    saved = []
+    for d in derivatives:
+        for saved_arg in d[key]:
+            if saved_arg['name'] in seen:
+                continue
+            seen.add(saved_arg['name'])
+            saved.append(saved_arg)
+    return saved
 
-    def split_name_params(prototype):
-        name, params = re.match('(\w+)\((.*)\)', prototype).groups()
-        return name, params.split(', ')
 
-    def get_signature(option):
-        arguments = option['python_arguments']
-        arg_types = [arg['type'] for arg in arguments]
-        if option['aten'] is not None:
-            call_args = split_name_params(option['aten'])[1]
-            arg_indices = {arg['name']: i for i, arg in enumerate(arguments)}
+def to_camel_case(name):
+    return ''.join([p.title() for p in name.split('_')])
 
-            def get_type(arg_name):
-                if arg_name not in arg_indices:
-                    # if the name is not an argument, assume it's a literal
-                    # number, with type 'Scalar'
-                    return 'Scalar'
-                return arg_types[arg_indices[arg_name]]
 
-            arg_types = [get_type(arg_name) for arg_name in call_args]
-        return '{}({})'.format(option['name'], ', '.join(arg_types))
+def split_name_params(prototype):
+    name, params = re.match('(\w+)\((.*)\)', prototype).groups()
+    return name, params.split(', ')
+
+
+def load_derivatives(path, declarations_by_signature):
+    with open(path, 'r') as f:
+        definitions = yaml.load(f, Loader=Loader)
+
+    def canonical_declaration(declarations, name):
+        for declaration in declarations:
+            if declaration['name'] == name:
+                return declaration
+        return declarations[0]
 
     # Parse each entry from derivatives.yaml
-    options = []
+    autograd_functions = []
     for defn in definitions:
-        option = {}
         if '(' not in defn['name']:
             continue
+
         name, params = split_name_params(defn['name'])
-        num_tensor_inputs = 0
-        option['name'] = name
-        option['aten'] = defn.get('aten')
-        option['python_arguments'] = []
-        option['prototype'] = defn['name']  # with default
-        option['fallthrough'] = defn.get('fallthrough', False)
-        option['op'] = name[0].upper() + name[1:] + 'Backward'
-
-        arg_sizes_found = []
+        param_types = [p.split(' ')[0] for p in params if p != '*']
+        signature = '{}({})'.format(name, ', '.join(param_types))
+
+        declarations = declarations_by_signature[signature]
+        if len(declarations) == 0:
+            raise RuntimeError('no ATen declaration found for: {}'.format(signature))
+        canonical = canonical_declaration(declarations, name)
+
+        num_inputs = 0
         derivatives = []
-        for param in params:
-            if param == '' or param == '*':
+        for arg in canonical['arguments']:
+            if arg['name'] not in defn:
```

Comment:
does this mean that we skipped some of ATen arguments? when is it valid to do that?